### PR TITLE
[ContextParallel][FlexAttention] Prototype of supporting FlexAttention in Context Parallel

### PR DESCRIPTION
**Pull request by XilunWu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #158692

**Summary**
This PR adds an all-gather based FlexAttention and uses TorchFunctionMode to dispatch
`FlexAttentionHOP.__call__` to it.

This PR makes the following changes:

- add a user-facing API `create_cp_block_mask` for creating CP-specific `BlockMask`
which masks over the attention result of Q shard and KV global.
- add `_ContextParallelGlobalVars` to store all necessary global vars that CP FlexAttention
requires. `torch_function_mode` is critical to maintain singleton mode to avoid dynamo
recompilations.
- add a dispatch path for `FlexAttentionForwardHOP.__call__` (TorchFunctionMode dispatch
won't work correctly without this line)

What's not in this PR:
- QKV load balancing
- Test on other masking besides `causal_mask`.
- Support on small attention (i.e. qkv size is smaller than 128) because the block mask
rewrite function requires `Q_BLOCK_SIZE == KV_BLOCK_SIZE == 128`.

**Test**
`pytest test/distributed/tensor/test_attention.py -s -k test_ring_flex_attention`

**Followup**
1. create an issue to reproduce the error in `create_fw_bw_graph()` when trying to call `create_block_mask`
to re-write `block_mask` in `FlexAttentionHOP` dispatch in `TorchFunctionMode`.
2. Merge `_ContextParallelGlobalVars` and `_cp_options`.

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @Lucaskabela

### CODE REVIEW

File: test/distributed/tensor/test_attention.py [hunk 1]
```diff
@@ -437,5 +439,172 @@ def _test_ring_attention_custom_transformer(self, rotater: _RotateMethod) -> Non
...
+        )
+        expect_out.sum().backward()
+
+        # Prepare the required global vars for CP+Flex:
+        device_mesh = init_device_mesh(
+            device_type=self.device_type,
+            mesh_shape=(self.world_size,),
+            mesh_dim_names=("cp",),
+        )
+        # NOTE: call create_block_mask() within TorchFunctionMode would cause error in create_fw_bw_graph
```
Comment (drisspg) [inline test/distributed/tensor/test_attention.py]: You created an issue for this right?

File: test/distributed/tensor/test_attention.py [hunk 2]
```diff
@@ -437,5 +439,172 @@ def _test_ring_attention_custom_transformer(self, rotater: _RotateMethod) -> Non
...
+            buffer_seq_dims=[2, 2, 2],
+        ):
+            cp_q.requires_grad = True
+            cp_k.requires_grad = True
+            cp_v.requires_grad = True
+
+            # this is the block_mask created within the training step
+            # NOTE: flex_attention checks block_mask shape and input shape before
+            # calling into flex_attention_hop.
+            block_mask_post_sharding = create_block_mask(
```
Comment (drisspg) [inline test/distributed/tensor/test_attention.py]: having to recreate the block mask doesn't seem right, you if you know that the block-mask will work you need not recreate, or Ig uess maybe I am confused thatyou need to create teh first one?

File: torch/distributed/tensor/experimental/_attention.py [hunk 3]
```diff
@@ -1086,6 +1115,74 @@ def backward_hook(grad: torch.Tensor) -> None:
...
+            return blk_idx * block_size + local_blk_offset
+
+        return lambda b, h, q_idx, kv_idx: mask_mod(
+            b,
+            h,
+            local_q_idx_to_q_idx(q_idx),
+            kv_idx,
+        )
+
+    def _create_block_mask(
```
Comment (drisspg) [inline torch/distributed/tensor/experimental/_attention.py]: Do we need this?

File: torch/distributed/tensor/experimental/_attention.py [hunk 4]
```diff
@@ -1139,6 +1283,36 @@ def __torch_function__(
         ) -> Any:
             kwargs = kwargs or {}
 
+            # special handler for flex_attention
+            if func == torch._higher_order_ops.flex_attention:
+                # raise NotImplementedError("flex_attention is not supported yet.")
```
Comment (drisspg) [inline torch/distributed/tensor/experimental/_attention.py]: nit: remove the comments

File: test/distributed/tensor/test_attention.py (hunk 1 above)
Comment (XilunWu) [reply test/distributed/tensor/test_attention.py]: Yes this was brought up in https://github.com/pytorch/pytorch/issues/157015 

let me create a sub-issue for this specifically.

File: test/distributed/tensor/test_attention.py (hunk 2 above)
Comment (XilunWu) [reply test/distributed/tensor/test_attention.py]: let me elaborate on this since it is quite important. This PR's CP + Flex impl needs 3 `BlockMask`:

1. block_mask on q_global, k_global, v_global. This provides the correct masking function.
2. block_mask on q_shard, k_shard, v_shard. This provides the correct shape to pass the shape check in `flex_attention` function (https://github.com/pytorch/pytorch/blob/6100ed457c9bf19dd80e0d53301c7bae691da8d3/torch/nn/attention/flex_attention.py#L1485-L1498 and https://github.com/pytorch/pytorch/blob/6100ed457c9bf19dd80e0d53301c7bae691da8d3/torch/nn/attention/flex_attention.py#L1527C9-L1542) when it's called on qkv shards.
3. block_mask on q_shard, k_global, v_global. This is the `BlockMask` passed to `FlexAttentionHOP` after all-gathering KV and it is computed from the first `block_mask`.

Here this `block_mask_post_sharding` is the second `block_mask` which is used to pass the shape check.

File: torch/distributed/tensor/experimental/_attention.py (hunk 3 above)
Comment (XilunWu) [reply torch/distributed/tensor/experimental/_attention.py]: good catch! I added this because I was planning to add `lru_cache` but actually this cache is not needed because I'm now requiring user to explicitly initialize the `BlockMask`s instead of implicitly `create_block_mask` in `FlexAttentionHOP` dispatch.

I do have another question: when should I compile `create_block_mask`?

File: torch/_dynamo/variables/higher_order_ops.py [hunk 5]
```diff
@@ -2793,6 +2793,11 @@ def call_function(
         args: "list[VariableTracker]",
         kwargs: "dict[str, VariableTracker]",
     ) -> "VariableTracker":
+        from .torch_function import can_dispatch_torch_function, dispatch_torch_function
+
+        if can_dispatch_torch_function(tx, args, kwargs):
+            return dispatch_torch_function(tx, self, args, kwargs)
+
```
Comment (zou3519) [inline torch/_dynamo/variables/higher_order_ops.py:2790]: seems reasonable to me, but (1) can we make sure that in eager-mode, flexattention has a torch_function and (2) can we test that in both eager and compile, flexattention torch_function gets dispatched?

Comment (XilunWu) [reply torch/_dynamo/variables/higher_order_ops.py:2790]: sounds good Richard, let me sync with @mlazos to discuss the merging of this change.

File: test/distributed/tensor/test_attention.py (hunk 2 above)
Comment (XilunWu) [reply test/distributed/tensor/test_attention.py]: sync with Driss offline: we can request users to explicitly call a new API `create_cp_block_mask()` that returns (3)rd `BlockMask` but the Q_LEN and K_LEN are modified to those in (2)nd, and reset to the right value right before passing to `FlexAttentionHOP.__call__` in `TorchFunctionMode`.

File: torch/distributed/tensor/experimental/_attention.py (hunk 3 above)
Comment (drisspg) [reply torch/distributed/tensor/experimental/_attention.py]: I typically create a global `create_block_mask_compiled` and use that one

File: test/distributed/tensor/test_attention.py (hunk 2 above)
Comment (fegin) [reply test/distributed/tensor/test_attention.py]: Just want to confirm, is the proposal adding `create_cp_block_mask` to the CP APIs?

File: test/distributed/tensor/test_attention.py [hunk 6]
```diff
@@ -437,5 +439,172 @@ def _test_ring_attention_custom_transformer(self, rotater: _RotateMethod) -> Non
...
+
+    @skip_if_lt_x_gpu(2)
+    @with_comms
+    def test_ring_flex_attention(self) -> None:
+        self.run_subtests(
+            {"qkv_size": [128 * self.world_size, 2048]},
+            self._test_ring_flex_attention,
+        )
+
+        # TODO: the current CP impl doesn't support small attentions (block_size < 128)
```
Comment (fegin) [inline test/distributed/tensor/test_attention.py]: Is this CP's limitation? Where is the limitation from?

File: torch/distributed/tensor/experimental/_attention.py [hunk 7]
```diff
@@ -66,6 +71,29 @@ class _ContextParallelOptions:
 _cp_options = _ContextParallelOptions()
 
 
+@dataclass
+class _ContextParallelGlobalVars:
```
Comment (fegin) [inline torch/distributed/tensor/experimental/_attention.py:94]: Since we have this dataclass, should we move `_dispatch_mode` and `_cp_options` to this dataclass?

File: test/distributed/tensor/test_attention.py (hunk 2 above)
Comment (XilunWu) [reply test/distributed/tensor/test_attention.py]: @fegin yes, because we really want to only create 1 `BlockMask` object and that's all. To reach this goal, the most straightforward solution is to have users call `create_cp_block_mask` explicitly. This API returns a `BlockMask` on q_shard, k_global, and v_global, but with `Q_LEN` and `KV_LEN` be q_shard/kv_shard shape, we rewrite the shape before actually using it in FlexAttention.

We can discuss how to improve this UX later but for now I suggest use this simple approach first.

File: test/distributed/tensor/test_attention.py (hunk 6 above)
Comment (XilunWu) [reply test/distributed/tensor/test_attention.py]: CP limitation in assuming Q_BLOCK_SIZE == KV_BLOCK_SIZE == 128, and small attention size will have those 2 sizes be 256. But it also makes sense that users won't use CP on small attention size.

File: torch/distributed/tensor/experimental/_attention.py (hunk 7 above)
Comment (XilunWu) [reply torch/distributed/tensor/experimental/_attention.py:94]: Yes we can merge them. I'll do that in a follow-up PR (this requires torchtitan change as well).

**Review by drisspg — APPROVED** [review]
